bpo-35113: Fix inspect.getsource to return correct source for inner classes

tirkarthi opened the pull request:
Previous approach used regex along with indentation of the match to return the source code. This caused below issues. This PR proposes usage of ast module to do correct parsing.  

**Issues**

1. When a class with same name is defined under different classes.
2. When a class definition is part of a string declaration.
3. When a class definition is part of multiline comment.

Examples of incorrect output can be found with https://bugs.python.org/issue35113. 

This PR also adds returning class decorators as part of the source code instead of just returning the class definition. 

**Inner class example :** 

```python
import inspect

class Foo:
    class Bar:
        def foo(self):
            pass

class Spam:
    class Bar:
        def bar(self):
            pass

print(inspect.getsource(Spam.Bar))
```

**On master**

```
$ ./python.exe foo.py
    class Bar:
        def foo(self):
            pass

```

**IPython**

`inspect.getsource` is also used in lot of tools like IPython and pdb which also return incorrect output due to this

```python
ipython
Python 3.6.4 (default, Mar 12 2018, 13:42:53)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.3.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import foo

In [2]: foo.Spam.Bar??
Init signature: foo.Spam.Bar()
Docstring:      <no docstring>
Source:
    class Bar:
        def foo(self):
            pass
File:           ~/stuff/python/backups/foo.py
Type:           type

$ python3 foo.py
> /Users/karthikeyansingaravelan/stuff/python/backups/foo.py(13)<module>()
-> if __name__ == "__main__":
(Pdb) pdb.getsourcelines(Spam.Bar)
(['    class Bar:\n', '        def foo(self):\n', '            pass\n'], 2)
```

**With PR**

```
$ ./python.exe foo.py
    class Bar:
        def bar(self):
            pass

```

**Class decorator example**

```python
import inspect

def foo(cls):
    return cls

@foo
@foo
class Foo:
    pass

print(inspect.getsource(Foo))
```

**On master**

```
$ ./python.exe foo.py
class Foo:
    pass

```

**With PR**

```
$ ./python.exe foo.py
@foo
@foo
class Foo:
    pass
```

<!-- issue-number: [bpo-35113](https://bugs.python.org/issue35113) -->
https://bugs.python.org/issue35113
<!-- /issue-number -->

tirkarthi committed:
Use ast module to find class definition

tirkarthi committed:
Add NEWS entry

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,58 @@ def findsource(object):
+
+        class ClassVisitor(ast.NodeVisitor):
+
+            def recursive(func):
```
This decorator is only used once, and it is not general. I think the code would be simpler if just adds corresponding code in `visit_ClassDef()`.

tirkarthi committed:
Fix class with multiple children and move decorator code to the method

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,57 @@ def findsource(object):
+                set_top_level_class = False
+
+                # Push the node on stack when there is a module level class
+                if not self.stack:
```
Why only if the stack empty? What about deeper nesting?
```python
class A:
    class B:
        class C:
            pass
```

tirkarthi replied:
I have refactored the code as per the comments. This was due to my stack management logic to add top level module and to push children class node inside the for loop. Since I have refactored the code to push and pop only at one place it's not needed. There is a similar test at https://github.com/python/cpython/pull/10307/files#diff-041789a617178e1a31ea7a1f86a5d2faR204 . Let me know if I need to add this also as a test.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,57 @@ def findsource(object):
+
+        if class_visitor.line_number is not None:
+            line_number = class_visitor.line_number
+            decorator_pattern = re.compile(r'^(\s*@)')
```
Just look at the first item of `decorator_list` in `visit_ClassDef`.

tirkarthi replied:
Thanks, I have handled the decorator code inside `visit_ClassDef` with `node.decorator_list` and it's much simpler now.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,57 @@ def findsource(object):
+
+            def __init__(self, source, qualname, *args, **kwargs):
+                self.stack = []
+                self.source = source
```
Is it used?

tirkarthi replied:
Not needed. Removed. Thanks.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,57 @@ def findsource(object):
+
+        class ClassVisitor(ast.NodeVisitor):
+
+            def __init__(self, source, qualname, *args, **kwargs):
```
Are args and kwargs used?

tirkarthi replied:
Not needed. Removed. Thanks.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,57 @@ def findsource(object):
+                    set_top_level_class = True
+                    self.stack.append(node.name)
+
+                qualname = '.'.join(self.stack)
```
Just inline it.

tirkarthi replied:
I feel this is more readable that I am joining the stack to form qualname than `if self.qualname == '.'join(self.stack)` . Let me know if it needs to be inlined.

serhiy-storchaka replied:
It is needed if use local variables from the outer scope. Then the condition will be `if qualname == '.'join(stack)`.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,57 @@ def findsource(object):
+
+                for child in ast.iter_child_nodes(node):
+                    if isinstance(child, ast.ClassDef):
+                        self.stack.append(child.name)
```
Remove it and just add to the name to the stack at the beginning of `visit_ClassDef`. This will make the code simpler.

tirkarthi replied:
Thanks, added it to the top and the function is much simpler to read than my original one :)

tirkarthi committed:
Fix PR comments

1. Use node.decorator_list to select decorators
2. Remove unwanted variables in ClassVisitor
3. Simplify stack management as per review

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,42 @@ def findsource(object):
+
+            def visit_ClassDef(self, node):
+                self.stack.append(node.name)
+
```
I think it would be clearer if remove empty lines in this method. Left empty lines only between methods and classes.

tirkarthi replied:
Removed empty lines. Thanks.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,42 @@ def findsource(object):
+
+        class ClassVisitor(ast.NodeVisitor):
+
+            def __init__(self, qualname):
```
Since the class is not used outside of `findsource()` you can simplify the code by removing `__init__`. Inside the visitor you can just use local variables from the outer scope. This will save several lines. Add `nonlocal line_number` because it is set inside the visitor.

tirkarthi replied:
I tried removing the constructor and using local variables which removes use of `self` as in `self.stack`. But since stack is changed in three functions now I have to declare `nonlocal` on three and make other changes. After conversion I think using class variables is more self-contained and I have kept them. Let me know if you need to change them to use local variables.

serhiy-storchaka replied:
No, you don't have to use `nonlocal` with a stack, because you don't assign to this variable in these methods. Only `line_number` needs it, at only in a single place. I prefer more compact code, but it is up to you using the construct and instance attributes.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,42 @@ def findsource(object):
+                    # decrement by one since lines starts with indexing by zero
+                    self.line_number = line_number - 1
+
+                for child in ast.iter_child_nodes(node):
```
Just call `self.generic_visit(node)`.

tirkarthi replied:
Changed. Thanks.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,42 @@ def findsource(object):
-            # less whitespace first
-            candidates.sort()
-            return lines, candidates[0][1]
+        import ast
```
Add the comment like in `_signature_fromstr()`.

tirkarthi replied:
Added a comment on lazy import of ast like `_signature_fromstr()`

tirkarthi committed:
Add test for nested functions and async calls

tirkarthi committed:
Fix pydoc test since comments are returned now correctly

tirkarthi committed:
Set event loop policy as None to fix environment related change

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,52 @@ def findsource(object):
+                self.stack.pop()
+                self.stack.pop()
+
+            def visit_AsyncFunctionDef(self, node):
```
Just `visit_AsyncFunctionDef = visit_FunctionDef`.

tirkarthi replied:
Thanks, I will add this. I will also refactor class to use local variables. My assumption was that `stack.append` requires stack variable to be nonlocal since it mutates `stack`. I was wrong on that part. Thanks for clearing up and I will change it as suggested to use variables from outer scope making it more compact 👍

serhiy-storchaka commented on Lib/test/inspect_fodder2.py:
```diff
@@ -187,24 +187,51 @@ class cls185:
+        class cls205:
+            pass
+    class cls207:
+        class cls208:
```
Maybe use `cls205` for testing classes with same name but different `__qualname__`?

tirkarthi replied:
Used `cls205` in the test. Thanks.

serhiy-storchaka commented on Lib/test/inspect_fodder2.py:
```diff
@@ -187,24 +187,51 @@ class cls185:
+# line 211
+def func212():
+    class cls213:
+        def func(self):
```
This method is not needed.

tirkarthi replied:
Removed. Thanks

serhiy-storchaka commented on Lib/test/inspect_fodder2.py:
```diff
@@ -187,24 +187,51 @@ class cls185:
+# line 225
+async def func226():
+    class cls227:
+        def func(self):
```
This method is not needed.

tirkarthi replied:
Removed. Thanks

serhiy-storchaka commented on Lib/test/test_inspect.py:
```diff
@@ -690,15 +691,27 @@ def test_nested_class_definition(self):
+        self.assertSourceEqual(asyncio.run(mod2.func226()), 227, 229)
+        self.assertSourceEqual(mod2.cls227, 233, 237)
+        self.assertSourceEqual(asyncio.run(mod2.cls227().func234()), 235, 236)
+        asyncio.set_event_loop_policy(None)
```
Use `addCleanup()`.

tirkarthi replied:
Seems this test class inherits from `unittest.Test` and `self.addCleanup(asyncio.set_event_loop_policy, None)` doesn't work. Should I use `asyncio.set_event_loop_policy(None)` or change the base class from which it's inherited?

tirkarthi replied:
I tried adding `class TestBuggyCases(GetSourceBase, test_utils.TestCase):` but it seems it expects all the test cases to setup an event loop and gives below error. Any other better way to do this?

```
======================================================================
ERROR: test_with_comment_instead_of_docstring (test.test_inspect.TestBuggyCases)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_asyncio/utils.py", line 533, in tearDown
    self.unpatch_get_running_loop()
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_asyncio/utils.py", line 525, in unpatch_get_running_loop
    events._get_running_loop = self._get_running_loop
AttributeError: 'TestBuggyCases' object has no attribute '_get_running_loop'
```

serhiy-storchaka replied:
Why it doesn't work? There is no `unittest.Test`, this class inherits from `unittest.TestCase`.

tirkarthi replied:
Sorry, it works. I think I was having test failures and tested it during that time causing confusion that it's an environment change. Added the change as part of c6e5e2c133069bcb45568b141636b6b78e2348d3 and verified locally.

```
./python.exe -m test --fail-env-changed test_inspect
Run tests sequentially
0:00:00 load avg: 2.57 [1/1] test_inspect

== Tests result: SUCCESS ==

1 test OK.

Total duration: 2 sec 345 ms
Tests result: SUCCESS
```

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst:
```diff
@@ -0,0 +1,3 @@
+:meth:`inspect.getsource` now returns correct source code for inner class
+with same name as module level class. class decorators are also returned as
+part of the source.
```
Add "Path by ...".

tirkarthi replied:
Added patch attribute. Thanks :)

tirkarthi committed:
Refactor visit_AsyncFunctionDef and tests

tirkarthi committed:
Refactor to use local variables and fix tests

tirkarthi committed:
Add patch attribution

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,45 @@ def findsource(object):
+        class_visitor.visit(tree)
+
+        if line_number is not None:
+            line_number = line_number
```
Not needed.

tirkarthi replied:
Removed. Thanks.

tirkarthi committed:
Use self.addCleanup for asyncio

serhiy-storchaka commented on Lib/test/test_inspect.py:
```diff
@@ -676,6 +677,41 @@ def test_getsource_on_method(self):
+        self.assertSourceEqual(asyncio.run(mod2.func225()), 226, 227)
+        self.assertSourceEqual(mod2.cls226, 231, 235)
+        self.assertSourceEqual(asyncio.run(mod2.cls226().func232()), 233, 234)
+        self.addCleanup(asyncio.set_event_loop_policy, None)
```
It should be called before the first `asyncio.run()`. The purpose of using `addCleanup` is that the clean up code should be executed if any code above fails.

tirkarthi replied:
Thanks I moved it up before first `asyncio.run` and tested the same.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -792,25 +792,45 @@ def findsource(object):
+        # it's not used for other than this part.
+        import ast
+
+        class ClassVisitor(ast.NodeVisitor):
```
`ClassVisitor` is not very informative name. It visits not only classes. Maybe name it `ClassFinder` or something like?

tirkarthi replied:
Renamed to `ClassFinder` and the corresponding variable below to `class_finder`. Thanks.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst:
```diff
@@ -0,0 +1,3 @@
+:meth:`inspect.getsource` now returns correct source code for inner class
+with same name as module level class. class decorators are also returned as
```
```suggestion
with same name as module level class. Class decorators are also returned as
```

tirkarthi replied:
Changed. Thanks.

serhiy-storchaka reviewed (commented):
Just few nitpicks.

But see also the comment on the tracker.

tirkarthi committed:
Rename ClassVisitor to ClassFinder and fix asyncio cleanup

matrixise commented on Lib/test/inspect_fodder2.py:
```diff
@@ -137,3 +137,99 @@ def func136():
+    return cls213
+
+# line 217
+class cls213:
```
should be `cls218`

tirkarthi replied:
This to check that the inner class which has the same name `cls213` inside `func212` doesn't conflict with this class `cls213`. Previously regex was used and hence there were cases where two classes with same name defined under different scopes might conflict with each other returning wrong results.

matrixise commented on Lib/test/inspect_fodder2.py:
```diff
@@ -137,3 +137,99 @@ def func136():
+    return cls226
+
+# line 230
+class cls226:
```
should be `cls231`

tirkarthi replied:
Same as my above comment but this is to check for `async def` related code.

This to check that the inner class which has the same name `cls226` inside `func225` doesn't conflict with this class `cls226`. Previously regex was used and hence there were cases where two classes with same name defined under different scopes might conflict with each other returning wrong results.

matrixise commented on Lib/inspect.py:
```diff
@@ -792,25 +792,44 @@ def findsource(object):
+                    else:
+                        line_number = node.lineno
+                    # decrement by one since lines starts with indexing by zero
+                    line_number = line_number - 1
```
why not "line_number -= 1" ?

tirkarthi replied:
No specific reason to be honest :) I will update it along with other review comments in the second round of review. Thanks.

matrixise reviewed (changes requested):
thanks for your contributions

tirkarthi committed:
Return first class inside conditional in case of multiple definitions. Remove decorator for class source.

tirkarthi committed:
Add docstring to make the test correct

tirkarthi committed:
Modify NEWS entry regarding decorators

tirkarthi committed:
Return decorators too for bpo-15856

1st1 commented on Lib/inspect.py:
```diff
@@ -801,25 +801,48 @@ def findsource(object):
+
+                    # decrement by one since lines starts with indexing by zero
+                    line_number -= 1
+                    raise StopIteration(line_number)
```
I don't like this way of interrupting the AST traversal.  Just declare a custom exception, raise it, and then catch it.

1st1 commented on Lib/inspect.py:
```diff
@@ -801,25 +801,48 @@ def findsource(object):
+            visit_AsyncFunctionDef = visit_FunctionDef
+
+            def visit_ClassDef(self, node):
+                nonlocal line_number
```
I think you should move this class to the top level, and instead of using a nonlocal name, use a proper property, e.g. `self.line_number`.

tirkarthi replied:
I am currently using the name `_ClassFinder` similar to (_ParameterKind) and `ClassFoundException` respectively. Please let me know if I need to change these names or move the location of these definitions in the file.

Thanks

1st1 commented on Lib/inspect.py:
```diff
@@ -801,25 +801,48 @@ def findsource(object):
-            return lines, candidates[0][1]
+        # Lazy import ast because it's relatively heavy and
+        # it's not used for other than this part.
+        import ast
```
I think you can just import it at the top level. The `inspect` module is a pretty heavy module already.

1st1 reviewed (changes requested):
Almost there. I've left a few nit comments.

tirkarthi committed:
Move ast and the class source code to top. Use proper Exception.

serhiy-storchaka reviewed (approved):
:+1:

I just have found a way to optimize the visitor, but it may complicate the implementation. Will try after merging this PR.
